### ISSUE TITLE
[arm][asm] LLVM IAS marks ARM function symbol as Thumb (st_value LSB=1) if .type %function occurs after a .thumb directive

### ISSUE THREAD

**Issue by nickdesaulniers** [llvm/llvm-project#211376]
When assembling ARM assembly code, LLVM Integrated Assembler (LLVM IAS) determines whether a %function symbol is
  Thumb (setting LSB of st_value to 1) based on the active assembler state at the moment the .type symbol,
  %function directive is parsed, rather than the ISA state active where the label (symbol:) was defined.

  If an ARM function label is defined under .arm, but its .type symbol, %function directive appears later in the
  file while .thumb mode is active (such as the ENTRY / ENDPROC macro pattern in Linux kernel assembly
  arch/arm/kernel/head.S), LLVM IAS erroneously marks the ARM function as a Thumb symbol by assigning it an odd
  address (st_value bit 0 = 1). GNU Assembler (GAS) correctly assigns it an even address (st_value bit 0 = 0).

  ### Minimal Reproducer (test.s)

```asm
.text
.syntax unified
.arm
.global secondary_startup_arm
secondary_startup_arm:
  nop

.thumb
.global secondary_startup
secondary_startup:
  nop

.type secondary_startup, %function
.type secondary_startup_arm, %function
```

  ### Commands to Reproduce

```sh
# Assemble with GNU Assembler (GAS)
arm-linux-gnueabihf-as -mthumb test.s -o test_gas.o

# Assemble with LLVM Integrated Assembler (LLVM IAS)
clang --target=arm-linux-gnueabihf -mthumb -c test.s -o test_clang.o

# Inspect symbol tables
llvm-readelf -s test_gas.o
llvm-readelf -s test_clang.o
```

  ### Symbol Table Output Comparison

  #### GAS Output (test_gas.o) — Correct:

```
    Symbol table '.symtab' contains 9 entries:
       Num:    Value  Size Type    Bind   Vis       Ndx Name
         ...
         7: 00000000     0 FUNC    GLOBAL DEFAULT     1 secondary_startup_arm
         8: 00000005     0 FUNC    GLOBAL DEFAULT     1 secondary_startup
```

  • secondary_startup_arm has Value = 00000000 (LSB = 0). GAS correctly identifies this as an ARM function.

  #### LLVM IAS Output (test_clang.o) — Incorrect:

```
    Symbol table '.symtab' contains 5 entries:
       Num:    Value  Size Type    Bind   Vis       Ndx Name
         ...
         3: 00000001     0 FUNC    GLOBAL DEFAULT     2 secondary_startup_arm
         4: 00000005     0 FUNC    GLOBAL DEFAULT     2 secondary_startup
```
  • secondary_startup_arm has Value = 00000001 (LSB = 1). LLVM IAS incorrectly marks this as a Thumb function
  because .type secondary_startup_arm, %function was evaluated after .thumb.

  ### Expected Behavior

  secondary_startup_arm was defined while .arm was active. Its symbol table entry (st_value) should be 00000000
  (even address, LSB=0).

  ### Actual Behavior

  LLVM IAS evaluates the .type symbol, %function directive using the current assembler state (.thumb), setting bit
  0 of st_value to 1 (odd address).

  ### Impact

  This breaks secondary CPU bringup in the Linux kernel on ARM32 platforms (such as MediaTek) when compiled with
  CONFIG_THUMB2_KERNEL=y and Clang. In Linux's arch/arm/kernel/head.S, ENDPROC(secondary_startup_arm) expands to
  .type secondary_startup_arm, %function at the end of the file where .thumb is active. In C code referencing
  __pa_symbol(secondary_startup_arm), the resulting physical address written to hardware jump registers has bit 0
  set to 1, causing secondary CPUs to crash when booting out of reset in ARM mode.

See also: https://lore.kernel.org/llvm/20260722172443.534251-1-akkun11.open@gmail.com/

**Comment by emomaxd** [comment]
Looking into this. I have a local patch and regression test, just doing a clean before/after check before sending a pr.
